tips/TIP-0045/tip-0045.md

```diff
@@ -0,0 +1,1610 @@
+
+<details>
+  <summary>Reference Unlock</summary>
+  <blockquote>References a previous unlock to support unlocking multiple inputs owned by the same address.</blockquote>
+</details>
+<table>
+  <tr>
+    <td>
+      <b>Name</b>
+    </td>
+    <td>
+      <b>Type</b>
+    </td>
+    <td>
+      <b>Description</b>
+    </td>
+  </tr>
+  <tr>
+    <td>Unlock Type</td>
+    <td>uint8</td>
+    <td>Set to <strong>value 1</strong> to denote a <i>Reference Unlock</i>.</td>
+  </tr>
+  <tr>
+    <td>Reference</td>
+    <td>uint16</td>
+    <td>Represents the index of a previous unlock.</td>
+  </tr>
+</table>
+
+##### Reference Unlock Syntactic Validation
+
+- The _Reference Unlock_ at index `i` must have `Reference < i` and the unlock at index `Reference` must be a _Signature
+  Unlock_.
+
+Example: Consider a <i>Transaction</i> containing the <i>UTXO Inputs</i> 0, 1 and 2, where 0 and 2 are both spending
+outputs belonging to the same Ed25519 address `A` and 1 is spending from a different address `B`. This results in the
+following structure of the <i>Unlocks</i> part:
+
+| Index | Unlock                                                                               |
+| ----- | ------------------------------------------------------------------------------------ |
+| 0     | A _Signature Unlock_ holding the Ed25519 signature for address `A`.                  |
+| 1     | A _Signature Unlock_ holding the Ed25519 signature for address `B`.                  |
+| 2     | A _Reference Unlock_ which references 0, as both require the same signature for `A`. |
+
+##### Reference Unlock Semantic Validation
+
+- The _Signature Unlock_ at index `Reference`, to which the _Reference Unlock_ at index `i` points, must be a
+  [semantically valid signature unlock](#signature-unlock-semantic-validation) for the input at index `i`.
+
+### Work Score
+
+Let the work score of a transaction be defined as follows.
+
+- Let `Size Score` be `Work Score Parameters::DataByte * Transaction Size` where `Transaction Size` is the size of the
+  serialized _Transaction_.
+- Let `Input Score` be `Inputs Count * Work Score Parameters::Input`.
+- Let `Context Input Score` be `Context Inputs Count * Work Score Parameters::Context Input`.
+- Let `Output Score` be the sum of work scores of each output `Output` in `Outputs`:
+  - Let `Native Token Score` be `Work Score Parameters::Native Token * Native Tokens Count` if `Output` contains
+    `Native Tokens`, `0` otherwise.
+  - Let `Token Scheme Score` be `Work Score Parameters::Native Token` if `Output` contains a _Simple Token Scheme_, `0`
+    otherwise.
+  - Let `Staking Score` be `Work Score Parameters::Staking` if `Output` contains a _Staking Feature_, `0` otherwise.
+  - Let `Block Issuer Score` be `Work Score Parameters::Block Issuer` if `Output` contains a _Block Issuer Feature_, `0`
+    otherwise.
+  - Return
+    `Work Score Parameters::Output + Native Token Score + Token Scheme Score + Staking Score + Block Issuer Score`.
```

Comment:
just to make sure: adding the `Output` offset from the `WorkScoreParameters` is still needed for each output, right?

Reply:
Looking into this now, we're somehow not using the Output offset anymore in the code which is odd.

Reply:
Is there any update on this?

Reply:
We are most likely missing the parameter in our calculations and the intention is to have it, so the TIP is correct, but the node currently isn't. Tracked in https://github.com/iotaledger/iota-core/issues/564.